api/services/registry/server.go

```diff
@@ -106,3 +109,40 @@ func (s *Server) Proxy(stream Registry_ProxyServer) error {
 
 	return nil
 }
+
+// CopyWithDeadline copies data from a net.Conn using a read deadline. The
+// process will fail with a timeout error if no data is read for the defined
+// period.
+func CopyWithDeadline(conn net.Conn, w io.Writer) (int64, error) {
+	var (
+		t   = int64(0)
+		buf = make([]byte, 32*1024)
+	)
+	for {
+		err := conn.SetReadDeadline(time.Now().Add(readDeadline))
+		if err != nil {
+			return t, err
+		}
+		n, err := conn.Read(buf)
+		if err != nil {
+			if errors.Is(err, io.EOF) || isNetTimeout(err) {
+				break
+			}
+			return t, err
+		}
+		buf = buf[0:n]
```

Comment:
This seems problematic, because this will shrink the buf, and the next time conn.Read(buf) is called, it could be smaller.

Maybe it's best to delete this line, and do `w.Write(buf[0:n])` instead?

Reply:
Thanks. I missed that line when I made the previous change. Updated.